test/pfm/pfm_test.go

```diff
@@ -0,0 +1,237 @@
+}
+
+func NewTransferPaths(chain1, chain2, chain3 *ibctesting.TestChain) (*ibctesting.Path, *ibctesting.Path) {
+	path1 := ibctesting.NewPath(chain1, chain2)
+	path1.EndpointA.ChannelConfig.PortID = ibctesting.TransferPort
+	path1.EndpointB.ChannelConfig.PortID = ibctesting.TransferPort
+	path1.EndpointA.ChannelConfig.Version = types.Version
+	path1.EndpointB.ChannelConfig.Version = types.Version
+	path2 := ibctesting.NewPath(chain2, chain3)
+	path2.EndpointA.ChannelConfig.PortID = ibctesting.TransferPort
+	path2.EndpointB.ChannelConfig.PortID = ibctesting.TransferPort
+	path2.EndpointA.ChannelConfig.Version = types.Version
+	path2.EndpointB.ChannelConfig.Version = types.Version
+
+	return path1, path2
+}
+
+// TestPacketForwardMiddlewareTransfer sends a PFM transfer originating from Celestia to ChainA, then back to Celestia and finally to ChainB.
+// It verifies that Celestia forwards the packet successfully, the balance of the sender account on Celestia decreases by the amount sent,
+// and the balance of the receiver account on ChainB increases by the amount sent.
+func TestPacketForwardMiddlewareTransfer(t *testing.T) {
+	coordinator, chainA, celestia, chainB := SetupTest(t)
+	path1, path2 := NewTransferPaths(chainA, celestia, chainB)
+
+	coordinator.Setup(path1)
+	coordinator.Setup(path2)
+
+	celestiaApp := celestia.App.(*app.App)
+	originalCelestiaBalalance := celestiaApp.BankKeeper.GetBalance(celestia.GetContext(), celestia.SenderAccount.GetAddress(), sdk.DefaultBondDenom)
+
+	// Take half of the original balance
+	transferAmount := originalCelestiaBalalance.Amount.QuoRaw(2)
+	timeoutHeight := clienttypes.NewHeight(1, 300)
+	coinToSendToB := sdk.NewCoin(sdk.DefaultBondDenom, transferAmount)
+
+	// Forward the packet to ChainB
+	secondHopMetaData := &PacketMetadata{
+		Forward: &ForwardMetadata{
+			Receiver: chainB.SenderAccount.GetAddress().String(),
+			Channel:  path2.EndpointA.ChannelID,
+			Port:     path2.EndpointA.ChannelConfig.PortID,
+		},
+	}
+	nextBz, err := json.Marshal(secondHopMetaData)
+	require.NoError(t, err)
+	next := string(nextBz)
+
+	// Send it back to Celestia
+	firstHopMetaData := &PacketMetadata{
+		Forward: &ForwardMetadata{
+			Receiver: celestia.SenderAccount.GetAddress().String(),
+			Channel:  path1.EndpointA.ChannelID,
+			Port:     path1.EndpointA.ChannelConfig.PortID,
+			Next:     &next,
+		},
+	}
+	memo, err := json.Marshal(firstHopMetaData)
+	require.NoError(t, err)
+
+	// Transfer path: Celestia -> ChainA -> Celestia -> ChainB
+	msg := types.NewMsgTransfer(path1.EndpointB.ChannelConfig.PortID, path1.EndpointB.ChannelID, coinToSendToB, celestia.SenderAccount.GetAddress().String(), chainA.SenderAccount.GetAddress().String(), timeoutHeight, 0, string(memo))
+
+	res, err := celestia.SendMsgs(msg)
+	require.NoError(t, err)
+
+	packet, err := ibctesting.ParsePacketFromEvents(res.GetEvents())
+	require.NoError(t, err)
+
+	err = ForwardPacket([]*ibctesting.Path{path1, path1, path2}, packet)
```

Comment:
[question] why does path1 appear twice in: `]*ibctesting.Path{path1, path1, path2}` ?

Reply:
because we have three hops and two of them are between chains Celestia and Chain A. Those chains are path1.

1st hop: Celestia -> Chain A (path1)
2nd hop: Chain A -> Celestia (path1)
3rd hop: Celestia -> Chain B (path2)